File: generated-code/codecraft/swift/Sources/TransGenTest/RepairProperties.swift
/// Entity's repair properties
public struct RepairProperties {
    /// Valid target entity types
    let validTargets: [EntityType]

    /// Health restored in one tick
    let power: Int32

    /// Read RepairProperties from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> RepairProperties {
        var validTargets: [EntityType]
        let validTargetsSize = stream.readInt32()
        validTargets = (0..<validTargetsSize).map{ _ in
            var validTargetsSize: EntityType
            validTargetsSize = EntityType.readFrom(stream)
            return validTargetsSize
        }
        var power: Int32
        power = stream.readInt32()
        return RepairProperties(validTargets: validTargets, power: power)
    }

    /// Write RepairProperties to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(Int32(validTargets.count))
        for validTargetsElement in validTargets {
            validTargetsElement.writeTo(stream)
        }
        stream.writeInt32(power)
    }
}

File: generated-code/haskell/codecraft/TcpReadWrite/src/Model/RepairProperties.hs
module Model.RepairProperties where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.EntityType (EntityType)

-- | Entity's repair properties
data RepairProperties = RepairProperties {
    -- | Valid target entity types
    validTargets :: [EntityType],
    -- | Health restored in one tick
    power :: Int32 }
    deriving Show

instance Trans RepairProperties where
    read = do
        validTargets <- Trans.read
        power <- Trans.read
        return RepairProperties {
            validTargets,
            power }

    write RepairProperties {
        validTargets,
        power } = do
            Trans.write validTargets
            Trans.write power